### PR TITLE
Convenience type rule

### PR DESCRIPTION
**Pull request by facumenzella**
### Why?

> This PR tackles https://github.com/nicklockwood/SwiftFormat/issues/733

### What?

This PR tries to tackle an [existing rule](https://realm.github.io/SwiftLint/convenience_type.html) from Swiftlint. Types used to host static members should be enums to avoid instantiation.

> I excluded the rule from many tests. I wasn't sure about updating them.

### CODE REVIEW

File: Sources/Rules.swift [hunk 1]
```diff
@@ -709,6 +709,55 @@ public struct _FormatRules {
...
+    public let convenienceType = FormatRule(
+        help: "Converts types used for hosting only static members into enums.",
+        options: []
+    ) { formatter in
+
+        func rangeHostsOnlyStaticMembers(start startIndex: Int, end endIndex: Int) -> Bool {
+            var j = startIndex
+            while j < endIndex, let token = formatter.token(at: j) {
+                // exit if there's a explicit init
+                if token == .keyword("init") {
```
Comment (calda) [inline Sources/Rules.swift:722]: It seems like this implementation will run into the same issues as https://github.com/nicklockwood/SwiftFormat/issues/728. This should probably only check declarations at the top level of the type definition.

Some examples that I think may get handled incorrectly right now:

```swift
struct Namespace {
    static func staticFunction() {
        func nestedFunction() { /* ... */ }
    }
}
```

```swift
struct Namespace {
    static func staticFunction() {
        struct NestedType {
            init() {}
        }
    }
}
```

```swift
struct Namespace {
    struct TypeNestedInNamespace {
        let foo: Int
        let bar: Int
    }
}
```

```swift
struct Namespace {
    struct NestedNamespace {
       static let foo: Int
       static let bar: Int
    }
}
```

You could potentially use `parseDeclarations()` to get the list of top-level declarations for each type.

File: Rules.md [hunk 2]
```diff
@@ -9,6 +9,7 @@
 * [braces](#braces)
 * [consecutiveBlankLines](#consecutiveBlankLines)
 * [consecutiveSpaces](#consecutiveSpaces)
+* [convenienceType](#convenienceType)
```
Comment (calda) [inline Rules.md:12]: I know the SwiftLint rule is also called `convenienceType`, but I'm not really sure what "convenience" means in this context. I've always heard caseless enums referred to as [namespaces](https://khanlou.com/2016/06/easy-namespacing-in-swift/). I think a name like `enumNamespaces` could make sense here.

File: Tests/RulesTests+Braces.swift [hunk 3]
```diff
@@ -57,7 +57,7 @@ extension RulesTests {
             // foo
         }
         """
-        testFormatting(for: input, rule: FormatRules.braces)
+        testFormatting(for: input, rule: FormatRules.braces, exclude: ["convenienceType"])
```
Comment (calda) [inline Tests/RulesTests+Braces.swift:60]: I noticed that the SwiftLint rule doesn't trigger for empty types like `class Foo {}`. That may be a nice addition to this implementation (and would avoid the need to exclude this rule from so many existing tests)

File: Sources/Rules.swift (hunk 1 above)
Comment (facumenzella) [reply Sources/Rules.swift:722]: Thanks for the review!!
I'll add some tests for that and fix them.

File: Snapshots/Layout/Layout/LayoutManaged.swift [hunk 4]
```diff
@@ -3,7 +3,7 @@
 import Foundation
 
 /// A protocol for view-type classes that can be configured using Layout
-@objc protocol LayoutConfigurable: class {
+@objc protocol LayoutConfigurable: enum {
```
Comment (nicklockwood) [inline Snapshots/Layout/Layout/LayoutManaged.swift:6]: @facumenzella I didn't spot this before, but this looks like a bug. I'd guess the rule wrongly identified this as a class whereas in fact it's a protocol.

Comment (facumenzella) [reply Snapshots/Layout/Layout/LayoutManaged.swift:6]: Oh I completely missed this one. I thought it was under control. I'll take a look

Comment (nicklockwood) [reply Snapshots/Layout/Layout/LayoutManaged.swift:6]: @facumenzella don't worry, I checked and the rule itself is fine (you added a unit test for this case) - I think you must've just forgotten to roll back this change to the snapshot tests.

**Pull request merged**